text/1011-query-rewrite.rst

```diff
@@ -0,0 +1,322 @@
+filtering policy is defined for a type, then all elements become invisible by
+default and must be made visible by at least one ``PERMIT`` policy.
+
+The filtering actions are defined as follows:
+
+- ``PERMIT`` -- add all objects, for which the expression evaluates to a
+  set containing at least one ``true`` value, to the set of visible objects.
+  ``PERMIT`` policies are combined using the ``OR`` operator.
+
+- ``RESTRICT TO`` -- restrict the set of visible objects to only objects for
+  which the expression evaluates to a set containing at least one ``true``
+  value.  ``RESTRICT`` policies are combined using the ``AND`` operator and
+  are applied *after* all ``PERMIT`` policies.
+
+The ``CHECK`` policy action is a *validation* action and is applied to
+``INSERT`` and ``UPDATE`` statements.  An object modification will be
+allowed only if the ``<element-expr>`` returns a set containing at least one
+``true`` value, otherwise an error will be raised and the statement execution
+aborted.  Note that in ``UPDATE`` the expression is applied to the object will
+all updates applied, not the original object.
+
+The ``DO AFTER`` and ``DO ON COMMIT`` are *trigger* actions and are applied
+to the result of a statement *as a whole*.  The result set is available in
+the ``<set-expr>`` expression as ``__result__``.  ``DO AFTER`` action
+is applied directly onto the result of an affected statement, whereas
+``DO ON COMMIT`` is applied shortly after the transaction containing the
+affected statement has been successfully committed.  Thus, ``DO AFTER`` is
+suitable for actions that are themselves transactional in nature and can be
+rolled back together with the relevant statement, such as writing into an
+audit log set.  ``DO ON COMMIT``, on the other hand, is designed for actions
+that generate non-transactional external side effects, such as emitting an
+event on an external message bus.  Unlike other actions, errors raised in
+``DO ON COMMIT`` action do not affect the original statement and will only
+be logged.
+
+Allowed statement/action combination are as follows:
+
++--------------+--------+--------+--------+--------+
+|              | SELECT | INSERT | UPDATE | DELETE |
++--------------+--------+--------+--------+--------+
+| PERMIT       |  yes   |   n/a  |  yes   |  yes   |
++--------------+--------+--------+--------+--------+
+| RESTRICT     |  yes   |   n/a  |  yes   |  yes   |
++--------------+--------+--------+--------+--------+
+| CHECK        |  n/a   |   yes  |  yes   |  n/a   |
++--------------+--------+--------+--------+--------+
+| DO AFTER     |  yes   |   yes  |  yes   |  yes   |
++--------------+--------+--------+--------+--------+
+| DO ON COMMIT |  n/a   |   yes  |  yes   |  yes   |
++--------------+--------+--------+--------+--------+
+
+It is an error to specify a policy action for a statement that does not
+support it.
+
+
+ALTER POLICY
+------------
+
+Alter the definition of a query rewrite policy.
```

Comment:
Which policy gets altered? It looks like we can have multiple policies on the same action. Are policies keyed on something?

Reply:
They are keyed on the definition verbatim, like constraints and indexes. See "open questions" comment.

Reply:
Ah, right. Used to get "open questions" section in RFC, but nevermind.